### PR TITLE
Avoid using globals for JSON RPC.

### PR DESCRIPTION
**Pull request by artek-koltun**
In this change JSON RPC related global variables were moved into a dedicated structure. It makes design flexible and testable. In addition tests are reworked to create a new JSON RPC structure per test. It makes the tests hermetic and allows to run them in parallel.
Patch provides backward compatibility for other bridges til they reworked to use JSONRPC interface injection

### CODE REVIEW

**Review by glimchb — CHANGES_REQUESTED** [review]
please split into several commits, try to avoid mixing logical functionality in the same commit
looks like globals is one of them and re-writing testing infra  is a separate commit

File: pkg/server/jsonrpc.go [hunk 1]
```diff
@@ -17,16 +17,22 @@ import (
...
+type JSONRPC struct {
+	socket string
+	id     uint64
+}
+
+// NewJSONRPC creates a new instance of JsonRpc
+func NewJSONRPC(socketPath string) *JSONRPC {
+	return &JSONRPC{
+		socket: socketPath,
+	}
```
Comment (glimchb) [inline pkg/server/jsonrpc.go:30]: let's set id explicitly to 1 or 0 ?

**Comment by artek-koltun** [comment]
> please split into several commits, try to avoid mixing logical functionality in the same commit looks like
> 
>     * JSONRPC struct to remove globals is one
> 
>     * Main flags restructure is second
> 
>     * and re-writing testing infra  is third commit

split

File: pkg/server/jsonrpc.go (hunk 1 above)
Comment (artek-koltun) [reply pkg/server/jsonrpc.go:31]: Done

**Review by glimchb — CHANGES_REQUESTED** [review]
the code looks ok, take me time to go over all the changes...
one thing that jumps out is that you change `Call` interface, which will lead to changes to all the bridge implementations that we have now...

**Comment by artek-koltun** [comment]
> the code looks ok, take me time to go over all the changes... one thing that jumps out is that you change `Call` interface, which will lead to changes to all the bridge implementations that we have now...

Do the other bridges always use the latest version from this SPDK bridge? Is it a right approach? Should they always point to a specific version and updated only on demand?🤔
I provided the Call() as well as test functions for backward compatibility and marked them as deprecated since they use a global instance. Ran marvell and nvidia unit tests on the local machine successfully. However, the main procedures there should be updated to inject a json rpc instance. To reduce any further disruptions in other bridges, I also converted JSONRPC to an interface and provided methods to create frontend, backend and middleend Servers. All new changes are split to commits, so please have a look

**Comment by glimchb** [comment]
> Do the other bridges always use the latest version from this SPDK bridge? 

no. they always point to a specific version. and we have renovate bot to propose update PRs

**Comment by artek-koltun** [comment]
> > Do the other bridges always use the latest version from this SPDK bridge?
> 
> no. they always point to a specific version. and we have renovate bot to propose update PRs

so if that bot creates a PR to update opi-spdk-bridge version, I think, I can modify main functions of other bridges directly in those PRs to leverage JSONRPC interface instead of the default global server.Call.

**Comment by glimchb** [comment]
> so if that bot creates a PR to update opi-spdk-bridge version, I think, I can modify main functions of other bridges directly in those PRs to leverage JSONRPC interface instead of the default global server.Call.

yes, I like this approach better

**Comment by artek-koltun** [comment]
> > so if that bot creates a PR to update opi-spdk-bridge version, I think, I can modify main functions of other bridges directly in those PRs to leverage JSONRPC interface instead of the default global server.Call.
> 
> yes, I like this approach better

The code in the patch is updated to provide a default global JSONRPC with the servers created by means of NewServer(). This should provide the required backward compatibility. The plan is the following:
* Merge the updates to use NewServer() for other bridges:
https://github.com/opiproject/opi-intel-bridge/pull/8
https://github.com/opiproject/opi-marvell-bridge/pull/144
https://github.com/opiproject/opi-nvidia-bridge/pull/127
An update for AMD is not prepared since I am not allowed to create a fork
* Merge this patch
* Update other bridges to use JSONRPC injection
* Remove the rest of global variables from opi-spdk-bridge

**Comment by glimchb** [comment]
> The plan is the following:
> 
> * Merge the updates to use NewServer() for other bridges:
>   [Update opi spdk bridge package opi-intel-bridge#8](https://github.com/opiproject/opi-intel-bridge/pull/8)
>   [Use NewServer to create opi-spdk server instances. opi-marvell-bridge#144](https://github.com/opiproject/opi-marvell-bridge/pull/144)
>   [Use NewServer to create opi-spdk server instances. opi-nvidia-bridge#127](https://github.com/opiproject/opi-nvidia-bridge/pull/127)

merged

File: pkg/backend/backend.go [hunk 2]
```diff
@@ -25,11 +26,21 @@ type Server struct {
...
+	rpc server.JSONRPC
 }
 
 // NewServer creates initialized instance of BackEnd server
 func NewServer() *Server {
-	return &Server{}
+	return NewServerWithJSONRPC(server.DefaultJSONRPC)
+}
+
+// NewServerWithJSONRPC creates initialized instance of BackEnd server communicating
```
Comment (glimchb) [inline pkg/backend/backend.go:38]: what is the plan here ? deprecate on of them in next PR ? or you want to keep both... ?

File: pkg/backend/backend_test.go [hunk 3]
```diff
@@ -9,15 +9,69 @@ import (
...
+	opiSpdkServer *Server
+	client        *backendClient
+	ln            net.Listener
+	testSocket    string
+	ctx           context.Context
+	conn          *grpc.ClientConn
+	jsonRPC       server.JSONRPC
+}
+
+func (e *testEnv) Close() {
```
Comment (glimchb) [inline pkg/backend/backend_test.go:40]: can this be in utils ? is it shared ?

File: pkg/backend/backend_test.go [hunk 4]
```diff
@@ -9,15 +9,69 @@ import (
...
+	server.CloseListener(e.ln)
+	if err := os.RemoveAll(e.testSocket); err != nil {
+		log.Fatal(err)
+	}
+	server.CloseGrpcConnection(e.conn)
+}
+
+func createTestEnvironment(startSpdkServer bool, spdkResponses []string) *testEnv {
+	env := &testEnv{}
```
Comment (glimchb) [inline pkg/backend/backend_test.go:49]: can this be in utils ? is it shared ?

**Review by glimchb — APPROVED** [review]
looks good, next patch will clean up things

File: pkg/backend/backend.go (hunk 2 above)
Comment (artek-koltun) [reply pkg/backend/backend.go:38]: discussed in slack

File: pkg/backend/backend_test.go (hunk 3 above)
Comment (artek-koltun) [reply pkg/backend/backend_test.go:40]: not yet, if we move it in utils, we will have a circular dependency at the moment. We need to introduce a couple of interfaces to break it.

**Pull request merged**